### PR TITLE
perf(core): add TwoRelationsQueries.replaceWithDelta(); switch org-user PUT to use it

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Refs [LOG-13474](https://linear.app/silverhand/issue/LOG-13474). Third task in the Phase 0.5 performance milestone. Stacked on [#8818](https://github.com/logto-io/logto/pull/8818) (Perf-A — the `organization_user_relations (tenant_id, user_id)` index) which the new method's delta CTE depends on at scale.

### What changed

- `packages/core/src/utils/RelationQueries.ts` — adds a new opt-in method `replaceWithDelta(schema1Id, schema2Ids)` on `TwoRelationsQueries`. Computes added/removed id sets in a single CTE statement (Postgres supports `INSERT` and `DELETE` in `WITH`) and returns `{ added, removed }`. Preserves the `select ... for update` row lock on the Schema1 entity that `replace()` already takes. The existing `replace()` method is left untouched.
- `packages/core/src/routes/organization/user/index.ts` — switches `PUT /organizations/:id/users` to call `replaceWithDelta()` instead of `replace()`. This is the only call site that migrates in this PR.
- `packages/core/src/utils/RelationQueries.test.ts` — new file. 3 unit tests using a slonik mock pool that captures every query: verifies the row lock fires first, the delta CTE structure (`with` graph + `returning` + `array_agg`), and the empty-result return shape.
- `packages/integration-tests/src/tests/api/organization/organization-user.test.ts` — 5 new integration cases under `PUT /organizations/:id/users delta semantics`: no-op PUT, partial-overlap, empty body, no-overlap-from-empty, and a role-preservation regression test that fails on master and passes after this change.
- `.changeset/perf-relation-queries-replace-delta.md` — `@logto/core`patch.

### Expected result

- `PUT /organizations/:id/users` with a no-op body produces zero row writes against `organization_user_relations` (master: ~2N).
- A one-row delta on an N-row membership produces exactly one row write.
- Members whose membership survives a PUT keep their role assignments. Previously, the `DELETE WHERE org_id = X` in `replace()` cascaded through `organization_role_user_relations` and silently dropped every member's roles on every PUT — even no-op PUTs. New behavior is strictly more correct.
- API response shape is unchanged (`204 No Content`). `replaceWithDelta`'s return value is discarded by this caller; it is wired up for downstream consumption by Phase 1 ([LOG-13462](https://linear.app/silverhand/issue/LOG-13462)).
- The other nine `TwoRelationsQueries` subclasses (SSO connector relations, the four `application_user_consent_*` tables, `organization_role_scope_relations`, `organization_role_resource_scope_relations`, `organization_invitation_role_relations`, `organization_jit_roles`) continue to use `replace()` and see no behavior change.

### Reviewer notes

- `replaceWithDelta` was kept as a separate method rather than rewriting `replace()` in place because of the cascade-preservation behavior difference. Of the ten `TwoRelationsQueries` subclasses, only two sit upstream of FK cascades (`organization_user_relations` and `organization_application_relations`), and only one of those (`organization_user_relations`) is the membership-set hot path. Opt-in migration bounds the blast radius to call sites that explicitly want the new semantics, and lets reviewers reason about each migration on its own.
- The `NOT IN` form for the deleted CTE was rewritten to `NOT EXISTS` after a self-review surfaced its NULL-fragility. `next_set.id` from `unnest(varchar[])` can't be NULL in normal operation, but the rewrite is null-safe and a touch more index-friendly.
- The unit-test mock pool is hand-rolled because slonik's `createMockPool` only exposes a `query` override, not `transaction`. The mock records every query (including the ones inside `pool.transaction(cb)`), which is sufficient for asserting structural invariants without locking in line-by-line SQL.

## Testing

Unit tests, integration tests

## Checklist
- [x] `.changeset`
- [x] unit tests
- [x] integration tests
- [x] necessary TSDoc comments

### CODE REVIEW

File: packages/core/src/routes/organization/user/index.ts [hunk 1]
```diff
@@ -113,7 +113,8 @@ export default function userRoutes(
         });
       }
 
-      await organizations.relations.users.replace(id, userIds);
+      // Membership can grow large; use the delta variant to avoid rewriting all rows.
+      await organizations.relations.users.replaceWithDelta(id, userIds);
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/routes/organization/user/index.ts:117]: Real bug, but intentionally out of scope here — request-body dedup (and the corresponding quota-counting correction) is the entire scope of [LOG-13470 / PR #8817](https://github.com/logto-io/logto/pull/8817), which lands strictly before this one in the same project's stack. The dedup change there flows through to this branch on rebase, and the resulting PUT handler ends up calling `replaceWithDelta()` on the already-deduped `userIds`. Splitting them keeps Perf-C scoped to the base-class method addition and one call-site migration.

File: packages/core/src/utils/RelationQueries.test.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  let cursor = 0;
+  const pool = createMockPool({
+    query: async (sql, values) => {
+      // eslint-disable-next-line @silverhand/fp/no-mutating-methods
+      captured.push({ sql, values });
+      const row = responses[cursor] ?? {};
+      // eslint-disable-next-line @silverhand/fp/no-mutation
+      cursor += 1;
+      return createMockQueryResult([row]);
+    },
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/utils/RelationQueries.test.ts:39]: Good call — addressed in the latest amend. The capturing pool now throws when the response cursor outruns the provisioned responses, so any future refactor that issues an unexpected extra query will fail the test loudly instead of falling through to an empty-row default.

File: packages/integration-tests/src/tests/api/organization/organization-user.test.ts [hunk 3]
```diff
@@ -134,4 +134,124 @@ describe('organization user APIs', () => {
...
+          .map(({ id }) => id)
+          .slice()
+          .sort()
+      ).toEqual([userA.id, userB.id].slice().sort());
+    });
+
+    it('should preserve role assignments of members whose membership survives a PUT', async () => {
+      // This is the cascade-correctness regression test. Master's full-rewrite `replace()`
+      // ran `DELETE WHERE org_id = X` which cascaded to `organization_role_user_relations`,
+      // dropping every member's roles on every PUT — even no-op PUTs. The delta `replace()`
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/integration-tests/src/tests/api/organization/organization-user.test.ts]: Renamed in the latest amend — the test comment now says `replaceWithDelta()` to match the actual method being exercised.

**Review by wangsijie — APPROVED** [review]

**Review by gao-sun — APPROVED** [review]

File: .changeset/perf-relation-queries-replace-delta.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+---
+"@logto/core": minor
```
Comment (gao-sun) [inline .changeset/perf-relation-queries-replace-delta.md]: `patch` as there's no user-visible new feature?

Comment (simeng-li) [reply .changeset/perf-relation-queries-replace-delta.md]: updated

File: .changeset/perf-relation-queries-replace-delta.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+---
+"@logto/core": patch
+---
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply .changeset/perf-relation-queries-replace-delta.md:3]: PR body updated
